oracle_fdw.c

```diff
@@ -752,19 +752,37 @@ deparseLimit(PlannerInfo *root, struct OracleFdwState *fdwState, RelOptInfo *bas
 		{
 			char *limit_val;
 			char *offset_val;
+			int add_limit_clause = true;
+			ListCell   *lc;
 
-			appendStringInfoString(&limit_clause, " FETCH FIRST ");
-			limit_val = datumToString(((Const *) root->parse->limitCount)->constvalue, ((Const *) root->parse->limitCount)->consttype);
-			if (IsA(root->parse->limitOffset, Const) && !((Const *) root->parse->limitOffset)->constisnull)
+			/* we don't append the LIMIT clause if there is aggregate function used */
+			foreach (lc, root->processed_tlist)
 			{
-				offset_val = datumToString(((Const *) root->parse->limitOffset)->constvalue, ((Const *) root->parse->limitOffset)->consttype);
-				appendStringInfo(&limit_clause, "%s+%s", limit_val, offset_val);
+				Node *ptl = (Node *) lfirst(lc);
+				if (IsA(ptl, TargetEntry))
+				{
+					TargetEntry *tge = (TargetEntry *) ptl;
+					if (IsA(tge->expr, Aggref))
+						add_limit_clause = false;
+				}
```

Comment:
That is not good enough; it only checks if the top-level node is an aggregate. You have to check if there is an aggregate anywhere in the plan:

```
EXPLAIN (COSTS OFF) SELECT 12 - count(*) FROM typetest1 LIMIT 1;

 Limit
   ->  Aggregate
         ->  Foreign Scan on typetest1
               Oracle query: SELECT /*5d31d5dda80a4849b792294e4bd31b17*/ '1' FROM "TYPETEST1" r1 FETCH FIRST 1 ROWS ONLY
(4 rows)
```

Reply:
> That is not good enough; it only checks if the top-level node is an aggregate. You have to check if there is an aggregate anywhere in the plan:
> 
> ```
> EXPLAIN (COSTS OFF) SELECT 12 - count(*) FROM typetest1 LIMIT 1;
> 
>  Limit
>    ->  Aggregate
>          ->  Foreign Scan on typetest1
>                Oracle query: SELECT /*5d31d5dda80a4849b792294e4bd31b17*/ '1' FROM "TYPETEST1" r1 FETCH FIRST 1 ROWS ONLY
> (4 rows)
> ```

The main problem here is that aggregates must be pushed down, adding additional work to deparse OpExpr to handle this case adds development time that could be put into adding aggregate push down :-( As you see with this "ectoplasmic" patch I lack time to get it seriously and actually I was not aware that oracle_fdw doesn't push down GROUP BY and aggregates otherwise I would not have started this patch. I will have more time probably in June I will get back to this PR.

Reply:
I see. I guess that even if aggregates were pushed down, not all aggregates could be (`array_agg`), so this would still remain an issue.
I understand that you have limited capacity to work on this; so have I.
Thanks a lot for your contribution so far. If you don't beat me to it, I will try to pick up the ball where you left it.
It would be a pity to let this go to waste.